docs/design-proposal.md

```diff
@@ -0,0 +1,190 @@
+
+* Golang 1.22 or later
+* Make for builds (see the [Makefile](../Makefile) for build targets)
+* gRPC for client/server transport
+* mTLS for client/server encryption
+
+The TLS 1.3 cipher suites that will be used include:
+
+* tls.TLS_AES_128_GCM_SHA256
+* tls.TLS_AES_256_GCM_SHA384
+* tls.TLS_CHACHA20_POLY1305_SHA256
+
+`tls.TLS_AES_128_GCM_SHA256` and `tls.TLS_AES_256_GCM_SHA384` both use the AES 
+cipher in Galois/Counter Mode (GCM) for symmetric encryption. AES is a widely 
+used symmetric encryption algorithm that is considered safe and efficient. 128 
+and 256 represent key sizes, where a larger key size generally means more 
+security but at the cost of more computational power. The numbers in the 
+SHA256 and SHA384 are the lengths of hashes in bits. SHA-256 and SHA-384 are 
+examples of hash functions in the SHA-2 family and are widely used in 
+cryptographic applications and protocols. They have longer hash lengths 
+compared to SHA-1 which is considered insecure now.
+
+`tls.TLS_CHACHA20_POLY1305_SHA256` uses the ChaCha20 stream cipher and Poly1305
+for message authentication. Stream ciphers are an alternative to block ciphers
+like AES. ChaCha20 is a newer stream cipher that is designed to offer improved
+performance on devices without hardware AES acceleration, such as some mobile
+devices. Poly1305 is used for ensuring message integrity. A fundamental
+component of secure communications, message authentication codes (MACs) like
+Poly1305 help assure that a message has not been tampered with in transit.
+
+The use of these cipher suites in TLS 1.3 implies encryption which offers 
+confidentiality along with integrity and authenticity, which are crucial for 
+secure communication. 
+
+### ijrd
+
+The `ijrd` service provides APIs for uploading, listing and deleting root file
+systems on which jobs will execute, as well as start, stop, and stream the output
+of running job commands. The gRPC API is defined by the following protobuf
+definitions:
+
+* [job.proto](../proto/job.proto) - Job management APIs
+* [rootfs.proto](../proto/rootfs.proto) - Root file system management API
+
+```shell
+$ ijrd help
+Usage: ijrd
+
+Flags:
+  --server - the server name (should match that used in certificates)
+  --port   - the port to listin on (default: 50051)
+  --cert   - path to X509 certificates
+  --key    - path to X509 key file
+  --rootfs - folder in which to extract root file systems to
+```
+The `ijrd` command requires all flags but `--port` to be defined.
+
+### ijrctl
+
+The `irjctl` command line application will allow the user to manage jobs and
+root file systems on the `ijrd` service. It has two sub-commands: `job` and 
+`rootfs`:
+
+```shell
+$ ijrctl help
+Usage: ijrctl [job|rootfs]
+Use 'ijrctl [job|rootfs] help' for more information about a command.
+```
+
+Encryption cert and key paths will be set in a config file so the user doesn't have
+to specify them each call.
+
+#### ijrctl rootfs
```

Comment:
Replacing the rootfs in the context of a job is not a requirement of the challenge. You may choose to implement it, but it's not required for the interview.

Reply:
Gotcha. I'll remove the rootfs API when I update the PR.

To clarify, do you just want the host's /proc remounted so that the jobs only see their own processes, or something additional?

Reply:
Removed rootfs api.